engine/src/main/java/org/destinationsol/game/screens/MapScreen.java

```diff
@@ -15,14 +15,76 @@
 
 
 public class MapScreen extends SolUiBaseScreen {
+    private MapDrawer mapDrawer;
+    private UiTextButton cancelButton;
+    private UiTextButton zoomInButton;
+    private UiTextButton zoomOutButton;
+
+    @Override
+    public void onAdd(SolApplication solApplication) {
+        mapDrawer = solApplication.getGame().getMapDrawer();
+        mapDrawer.setToggled(true);
+        UiVerticalListLayout verticalListLayout = new UiVerticalListLayout();
+        cancelButton = new UiTextButton()
+                .setDisplayName("Cancel")
+                .setTriggerKey(solApplication.getOptions().getKeyClose())
+                .setOnReleaseAction(uiElement -> {
+                    mapDrawer.setToggled(false);
+                    SolApplication.changeScreen(solApplication.getGame().getScreens().mainGameScreen);
+                });
+        UiHeadlessButton alternateCancelKeyButton = new UiHeadlessButton()
+                .setTriggerKey(solApplication.getOptions().getKeyMap())
+                .setOnReleaseAction(uiElement -> {
+                   cancelButton.maybeFlashPressed(solApplication.getOptions().getKeyClose());
+                });
+        zoomOutButton = new UiTextButton()
+                .setDisplayName("Zoom out")
+                .setOnReleaseAction(uiElement -> {
+                    mapDrawer.changeZoom(false);
+                })
+                .setTriggerKey(solApplication.getOptions().getKeyZoomOut());
+        zoomInButton = new UiTextButton()
+                .setDisplayName("Zoom in")
+                .setOnReleaseAction(uiElement -> {
+                    mapDrawer.changeZoom(true);
+                })
+                .setTriggerKey(solApplication.getOptions().getKeyZoomIn());
+        verticalListLayout.addElement(cancelButton);
+        verticalListLayout.addElement(zoomInButton);
+        verticalListLayout.addElement(zoomOutButton);
+        verticalListLayout.addElement(alternateCancelKeyButton);
+        rootUiElement = new UiRelativeLayout().addElement(verticalListLayout, UI_POSITION_RIGHT, -UiTextButton.DEFAULT_BUTTON_WIDTH / 2, 0);
+        SolApplication.getInputManager().changeScreen(this);
+    }
+
+    @Override
+    public void updateCustom(SolApplication solApplication, SolInputManager.InputPointer[] inputPointers, boolean clickedOutside) {
+        Boolean scrolled = SolApplication.getInputManager().getScrolledUp();
+        if (scrolled != null) {
+            // If scrolling upwards
+            if (scrolled) {
+                zoomOutButton.maybeFlashPressed(solApplication.getOptions().getKeyZoomOut());
+            } else { // If scrolling downwards
+                zoomInButton.maybeFlashPressed(solApplication.getOptions().getKeyZoomIn());
+            }
+        }
+
+        ShipUiControl shipControl = solApplication.getGame().getScreens().mainGameScreen.shipControl;
+        if (shipControl instanceof ShipMouseControl) {
+            shipControl.update(solApplication, true);
```

Comment:
Why are you only updating for ShipMouseControl ?

Reply:
I found this code in the old system. I appears that it was implemented for the Android release as a way of controlling the ship from the map screen. I do not believe that I would have any effect with the other controls.